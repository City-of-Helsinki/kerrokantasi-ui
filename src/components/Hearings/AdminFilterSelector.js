import React from 'react';
import PropTypes from 'prop-types';
import Icon from '../../utils/Icon';
import {Nav, NavItem} from 'react-bootstrap';
import {FormattedMessage} from 'react-intl';

/**
 * [AdminFilterSelector description]
 * @param {object} props
 * @param {String} props.active
 * @param {function} props.onSelect
 * @param {Array<String|Object>} props.options
 */
const AdminFilterSelector = ({
  active,
  onSelect,
  options,
}) => {
  return (
    <Nav activeKey={active} bsStyle="pills" className="admin-filter-selector">
      {options.map((filter) => {
        const {list, iconName, formattedMessage, role} = filter;
        // NavItem role is button by default if filter.role doesnt exist
        return (
          <NavItem key={list} eventKey={list} onSelect={onSelect} role={role}>
            <Icon name={iconName} aria-hidden />
            <FormattedMessage id={formattedMessage}>{txt => txt}</FormattedMessage>
          </NavItem>
        );
      })}
    </Nav>
  );
};

AdminFilterSelector.propTypes = {
  active: PropTypes.string,
  onSelect: PropTypes.func,
  options: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.shape({
      list: PropTypes.string,
      formattedMessage: PropTypes.string,
    }))
  ]),
};

export default AdminFilterSelector;
