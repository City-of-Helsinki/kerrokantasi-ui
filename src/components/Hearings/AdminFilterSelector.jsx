import React from 'react';
import PropTypes from 'prop-types';
import { SelectionGroup, RadioButton } from 'hds-react';
import { FormattedMessage, useIntl } from 'react-intl';

/**
 * [AdminFilterSelector description]
 * @param {object} props
 * @param {String} props.active
 * @param {function} props.onSelect
 * @param {Array<String|Object>} props.options
 */
const AdminFilterSelector = ({ active, onSelect, options }) => {
  const intl = useIntl();

  return (
    <SelectionGroup
      direction='horizontal'
      className='admin-filter-selector'
      label={intl.formatMessage({ id: 'adminFilterSelectorLabel' })}
    >
      {options.map((filter) => {
        const { list, formattedMessage } = filter;
        return (
          <RadioButton
            key={list}
            id={`filter-${list}`}
            value={list}
            checked={active === list}
            onChange={() => onSelect(list)}
            label={<FormattedMessage id={formattedMessage}>{(txt) => txt}</FormattedMessage>}
          />
        );
      })}
    </SelectionGroup>
  );
};

AdminFilterSelector.propTypes = {
  active: PropTypes.string,
  onSelect: PropTypes.func,
  options: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(
      PropTypes.shape({
        list: PropTypes.string,
        formattedMessage: PropTypes.string,
      }),
    ),
  ]),
};

export default AdminFilterSelector;
