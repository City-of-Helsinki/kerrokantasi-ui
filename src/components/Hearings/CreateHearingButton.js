import React from 'react';
import PropTypes from 'prop-types';
import {FormattedMessage} from 'react-intl';
import {Button, ButtonToolbar} from 'react-bootstrap';
import Icon from '../../utils/Icon';
import Link from '../LinkWithLang';

const CreateHearingButton = ({to}) => (
  <div className="toolbar-bottom create-hearing-button-container">
    <ButtonToolbar className="actions pull-right">
      <Link to={to}>
        <Button bsStyle="success">
          <Icon name="plus"/>&nbsp;&nbsp;
          <FormattedMessage id="createHearing"/>
        </Button>
      </Link>
    </ButtonToolbar>
  </div>
);

CreateHearingButton.propTypes = {
  to: PropTypes.object.isRequired
};

export default CreateHearingButton;
