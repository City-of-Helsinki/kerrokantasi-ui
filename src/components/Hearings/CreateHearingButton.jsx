import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { Button } from 'hds-react';

import Icon from '../../utils/Icon';
import Link from '../LinkWithLang';

const CreateHearingButton = ({ to }) => (
  <div className='toolbar-bottom create-hearing-button-container'>
    <div className='actions pull-right'>
      <Link to={to}>
        <Button className='kerrokantasi-btn success'>
          <Icon name='plus' />
          &nbsp;&nbsp;
          <FormattedMessage id='createHearing' />
        </Button>
      </Link>
    </div>
  </div>
);

CreateHearingButton.propTypes = {
  to: PropTypes.object.isRequired,
};

export default CreateHearingButton;
