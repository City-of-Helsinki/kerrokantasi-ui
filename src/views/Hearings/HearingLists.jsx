const HearingLists = {
  ALL: {
    list: 'allHearings',
    isFetching: false,
    params: {},
    formattedMessage: 'allHearings',
    iconName: 'globe',
  },
  OPEN: {
    list: 'openHearings',
    isFetching: false,
    params: {},
    formattedMessage: 'openHearings',
    iconName: 'commenting-o',
  },
  PUBLISHED: {
    list: 'publishedHearings',
    isFetching: false,
    params: { published: 'True' },
    formattedMessage: 'publishedHearings',
    iconName: 'eye',
  },
  QUEUE: {
    list: 'publishingQueueHearings',
    isFetching: false,
    params: { published: 'True' },
    formattedMessage: 'publishingQueue',
    iconName: 'calendar-check-o',
  },
  DRAFTS: {
    list: 'draftHearings',
    isFetching: false,
    params: { published: 'False' },
    formattedMessage: 'drafts',
    iconName: 'pencil-square-o',
  },
  OWN: {
    list: 'ownHearings',
    isFetching: false,
    params: {},
    formattedMessage: 'ownHearings',
    iconName: 'user',
    role: 'link',
  },
};

export default HearingLists;

export const AdminFilters = [HearingLists.PUBLISHED, HearingLists.QUEUE, HearingLists.DRAFTS, HearingLists.OWN];