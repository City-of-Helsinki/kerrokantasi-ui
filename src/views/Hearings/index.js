// TODO: remove this disable once https://github.com/yannickcr/eslint-plugin-react/pull/1628 lands
/* eslint-disable react/no-unused-prop-types */
/* eslint-disable react/prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { injectIntl, intlShape, FormattedMessage } from 'react-intl';
import Helmet from 'react-helmet';
import { Col, Row } from 'react-bootstrap';
import { get, find, includes } from 'lodash';
import isEmpty from 'lodash/isEmpty';
import { withRouter } from 'react-router-dom';
import * as Actions from '../../actions';
import { isAdmin } from '../../utils/user';
import WrappedHearingList from '../../components/HearingList';
import LoadSpinner from '../../components/LoadSpinner';
import CreateHearingButton from '../../components/Hearings/CreateHearingButton';
import AdminFilterSelector from '../../components/Hearings/AdminFilterSelector';
import { hearingShape, labelShape, userShape } from '../../types';
import getAttr from '../../utils/getAttr';
import { parseQuery, stringifyQuery } from '../../utils/urlQuery';
import {getUser} from '../../selectors/user';

const now = () => new Date().toISOString();

const HearingLists = {
  ALL: {
    list: 'allHearings',
    params: {},
    formattedMessage: 'allHearings',
    iconName: 'globe'
  },
  OPEN: {
    list: 'openHearings',
    params: {},
    formattedMessage: 'openHearings',
    iconName: 'commenting-o'
  },
  PUBLISHED: {
    list: 'publishedHearings',
    params: { published: 'True' },
    formattedMessage: 'publishedHearings',
    iconName: 'eye'
  },
  QUEUE: {
    list: 'publishingQueueHearings',
    params: { published: 'True' },
    formattedMessage: 'publishingQueue',
    iconName: 'calendar-check-o'
  },
  DRAFTS: {
    list: 'draftHearings',
    params: { published: 'False' },
    formattedMessage: 'drafts',
    iconName: 'pencil-square-o'
  },
  OWN: {
    list: 'ownHearings',
    params: {},
    formattedMessage: 'ownHearings',
    iconName: 'user',
    role: 'link'
  }
};

const AdminFilters = [HearingLists.PUBLISHED, HearingLists.QUEUE, HearingLists.DRAFTS, HearingLists.OWN];

const getHearingListParams = listName => {
  const defaultParams = { include: 'geojson' };
  const hearingList = find(HearingLists, ({ list }) => list === listName);
  const hearingListParams = get(hearingList, 'params', {});

  const params = Object.assign({}, defaultParams, hearingListParams);

  switch (listName) {
    case HearingLists.PUBLISHED.list:
      params.open_at_lte = now();
      break;
    case HearingLists.QUEUE.list:
      params.open_at_gt = now();
      break;
    default:
    // Don't edit params further
  }
  return params;
};

export class Hearings extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      sortBy: '-created_at',
      adminFilter: isAdmin(props.user) ? AdminFilters[0].list : null,
      showOnlyOpen: false,
    };

    this.handleSearch = this.handleSearch.bind(this);
    this.handleSelectLabels = this.handleSelectLabels.bind(this);
    this.handleSort = this.handleSort.bind(this);
    this.setAdminFilter = this.setAdminFilter.bind(this);
    this.toggleShowOnlyOpen = this.toggleShowOnlyOpen.bind(this);
  }

  componentDidMount() {
    const { fetchLabels } = this.props;
    fetchLabels();
    // Hearing List is fetched when labels are available -> componentWillReceiveProps
  }

  componentWillReceiveProps(nextProps) {
    const { user, location, match: {params: {tab}}, labels } = this.props;
    const { adminFilter } = this.state;
    const shouldSetAdminFilter = isAdmin(nextProps.user) && (!user || !adminFilter);
    const shouldNullAdminFilter = isAdmin(user) && !nextProps.user;
    const shouldFetchHearings = labels && ((
      (!this.props.labels.length && nextProps.labels.length) ||
      (nextProps.labels.length && location.search !== nextProps.location.search) ||
      (!this.props.user && nextProps.user) ||
      (this.props.user && !nextProps.user)) || nextProps.match.params.tab !== tab);

    if (shouldSetAdminFilter) {
      this.setAdminFilter(AdminFilters[0].list);
    }

    if (shouldNullAdminFilter) {
      this.setAdminFilter(null);
    }

    if (shouldFetchHearings) {
      this.fetchHearingList(nextProps);
    }
  }

  getHearings() {
    const { hearingLists } = this.props;
    const hearingListKey = this.getHearingListName();
    return get(hearingLists, [hearingListKey, 'data'], null);
  }

  getHearingsCount() {
    const { hearingLists } = this.props;
    const list = this.getHearingListName();
    return get(hearingLists, [list, 'count'], 0);
  }

  getHearingListName() {
    const { user } = this.props;
    const { adminFilter } = this.state;

    return isAdmin(user) ? adminFilter : HearingLists.ALL.list;
  }

  getIsLoading() {
    const { hearingLists } = this.props;
    const name = this.getHearingListName();
    return get(hearingLists, [name, 'isFetching'], true);
  }

  static getSearchParams(props) {
    const { location } = props;
    const params = {};

    if (parseQuery(location.search).search) Object.assign(params, {title: parseQuery(location.search).search});
    if (parseQuery(location.search).label) {
      Object.assign(params, {label: Hearings.getLabelsFromQuery(parseQuery(location.search).label.toString())});
    }
    return params;
  }

  setAdminFilter(filter) {
    if (filter === 'ownHearings') {
      this.forwardToUserHearings();
    } else {
      this.setState({ adminFilter: filter }, () => this.fetchHearingList());
    }
  }

  forwardToUserHearings() {
    const {history, location} = this.props;
    const searchParams = parseQuery(location.search);
    history.push({
      pathname: '/user-hearings',
      search: stringifyQuery(searchParams)
    });
  }

  fetchHearingList(props = this.props) {
    const {fetchInitialHearingList, fetchAllHearings, match: {params: {tab}}} = props;
    const { sortBy } = this.state;
    const list = this.getHearingListName();
    const params = Object.assign({}, getHearingListParams(list), { ordering: sortBy }, Hearings.getSearchParams(props));

    if (tab === 'map') {
      fetchAllHearings(list, params);
    } else {
      fetchInitialHearingList(list, params);
    }
  }

  static getLabelsFromQuery = (labelsInQuery = []) => {
    if (Array.isArray(labelsInQuery)) return labelsInQuery;

    return [labelsInQuery];
  };

  handleSearch(searchTitle, force = false) {
    const { history, location } = this.props;
    const searchParams = parseQuery(location.search);
    const searchParamsEmpty = isEmpty(searchParams.search);
    const searchPhraseEmpty = isEmpty(searchTitle);
    const searchPhraseUpdated = searchParams.search !== searchTitle;

    // Don't do anything if:
    // search is not forced
    // AND
    // current search params are the same as the inputted search
    // OR
    // both search params and currently inputted search are empty
    if (!force && ((searchParams.search === searchTitle) || (searchParamsEmpty && searchPhraseEmpty))) {
      return;
    }

    if (searchPhraseEmpty) {
      delete searchParams.search;
    } else {
      searchParams.search = searchTitle;
    }

    if (searchPhraseUpdated || force) {
      history.push({
        path: location.path,
        search: stringifyQuery(searchParams),
      });
    }
  }

  handleSelectLabels(labels) {
    const { history, location } = this.props;
    const searchParams = parseQuery(location.search);
    searchParams.label = labels.map(({ id }) => id);
    history.push({
      path: location.pathname,
      search: stringifyQuery(searchParams)
    });
  }

  handleSort(sortBy) {
    this.setState(
      () => ({ sortBy }),
      () => {
        this.fetchHearingList();
      },
    );
  }

  toggleShowOnlyOpen() {
    this.setState(({ showOnlyOpen }) => ({ showOnlyOpen: !showOnlyOpen }));
  }

  handleReachBottom = () => {
    const {fetchMoreHearings, hearingLists} = this.props;
    const list = this.getHearingListName();

    if (
      hearingLists[list] &&
      hearingLists[list].count > hearingLists[list].data.length &&
      typeof hearingLists[list].next === 'string' &&
      !hearingLists[list].isLoading
    ) {
      fetchMoreHearings(list);
    }
  }

  render() {
    const {
      history,
      intl: { formatMessage },
      labels,
      language,
      match: { params: { tab } },
      location,
      user,
    } = this.props;
    const labelsInQuery = Array.isArray(parseQuery(location.search).label)
      ? parseQuery(location.search).label
      : [parseQuery(location.search).label];
    const selectedLabels = labels
      .filter(label => includes(labelsInQuery, label.id.toString())).map(label => getAttr(label.label, language));
    const searchTitle = parseQuery(location.search).search;
    const { showOnlyOpen } = this.state;
    const hearings = this.getHearings();
    const hearingCount = this.getHearingsCount();

    if (user && user.isFetching) {
      return <LoadSpinner />;
    }
    return (
      <div className="hearings">
        <section className="page-section page-section--all-hearings-header">
          <div className="container">
            <Row>
              <Col md={10} mdPush={1}>
                <Helmet title={formatMessage({ id: 'allHearings' })} />
                <FormattedMessage id="allHearings">
                  {allHearings => (
                    <h1
                      className="page-title"
                      // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
                      tabIndex="0"
                      aria-label={allHearings}
                      id="allHearingsPageTitle"
                    >
                      {allHearings}
                    </h1>
                  )}
                </FormattedMessage>
                {isAdmin(user) &&
                  <AdminFilterSelector
                  onSelect={this.setAdminFilter}
                  options={AdminFilters}
                  active={this.getHearingListName()}
                  />
                }
                {isAdmin(user) && <CreateHearingButton to={{path: '/hearing/new'}} />}
              </Col>
            </Row>
          </div>
        </section>
        {!isEmpty(labels) ? <WrappedHearingList
          hearings={hearings}
          hearingCount={hearingCount}
          selectedLabels={selectedLabels ? [].concat(selectedLabels) : []}
          searchPhrase={searchTitle}
          isLoading={this.getIsLoading()}
          labels={labels}
          showOnlyOpen={showOnlyOpen}
          handleSort={this.handleSort}
          handleSearch={this.handleSearch}
          handleSelectLabels={this.handleSelectLabels}
          toggleShowOnlyOpen={this.toggleShowOnlyOpen}
          language={language}
          tab={tab}
          intl={this.props.intl}
          handleReachBottom={this.handleReachBottom}
          onTabChange={value => {
            const url = `/hearings/${value}`;
            history.push({
              pathname: url,
              search: location.search,
            });
          }}
        /> : <LoadSpinner />}
      </div>
    );
  }
}

Hearings.propTypes = {
  fetchInitialHearingList: PropTypes.func,
  fetchMoreHearings: PropTypes.func,
  fetchLabels: PropTypes.func,
  hearingLists: PropTypes.objectOf(
    PropTypes.shape({
      isFetching: PropTypes.boolean,
      data: PropTypes.arrayOf(hearingShape),
    }),
  ),
  history: PropTypes.shape({
    push: PropTypes.func,
  }),
  intl: intlShape.isRequired,
  labels: PropTypes.arrayOf(labelShape),
  language: PropTypes.string,
  location: PropTypes.shape({
    search: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
  }),
  match: PropTypes.shape({
    params: PropTypes.shape({
      tab: PropTypes.string,
    }),
  }),
  user: PropTypes.shape({
    data: userShape,
    isFetching: PropTypes.boolean,
  }),
};

const mapStateToProps = state => ({
  hearingLists: state.hearingLists,
  labels: state.labels.data,
  language: state.language,
  user: getUser(state), // Using this on purpose without selector
});

const mapDispatchToProps = dispatch => ({
  fetchInitialHearingList: (listName, params) =>
    dispatch(Actions.fetchInitialHearingList(listName, '/v1/hearing/', params)),
  fetchAllHearings: (list, params) => dispatch(Actions.fetchHearingList(list, '/v1/hearing', params)),
  fetchMoreHearings: (listName) => dispatch(Actions.fetchMoreHearings(listName)),
  fetchLabels: () => dispatch(Actions.fetchLabels()),
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(injectIntl(Hearings)));
