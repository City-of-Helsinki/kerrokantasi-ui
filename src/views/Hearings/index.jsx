/* eslint-disable camelcase */
// TODO: remove this disable once https://github.com/yannickcr/eslint-plugin-react/pull/1628 lands
/* eslint-disable react/no-unused-prop-types */
/* eslint-disable react/prop-types */
import React, { useCallback, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { injectIntl, FormattedMessage } from 'react-intl';
import Helmet from 'react-helmet';
import { Col, Row } from 'react-bootstrap';
import { get, find, includes } from 'lodash';
import isEmpty from 'lodash/isEmpty';
import { useLocation, useNavigate, useParams } from 'react-router-dom';

import * as Actions from '../../actions';
import { isAdmin } from '../../utils/user';
import WrappedHearingList from '../../components/HearingList';
import LoadSpinner from '../../components/LoadSpinner';
import CreateHearingButton from '../../components/Hearings/CreateHearingButton';
import AdminFilterSelector from '../../components/Hearings/AdminFilterSelector';
import { hearingShape, labelShape, userShape } from '../../types';
import getAttr from '../../utils/getAttr';
import { parseQuery, stringifyQuery } from '../../utils/urlQuery';
import getUser from '../../selectors/user';

const now = () => new Date().toISOString();

const HearingLists = {
  ALL: {
    list: 'allHearings',
    isFetching: false,
    params: {},
    formattedMessage: 'allHearings',
    iconName: 'globe',
  },
  OPEN: {
    list: 'openHearings',
    isFetching: false,
    params: {},
    formattedMessage: 'openHearings',
    iconName: 'commenting-o',
  },
  PUBLISHED: {
    list: 'publishedHearings',
    isFetching: false,
    params: { published: 'True' },
    formattedMessage: 'publishedHearings',
    iconName: 'eye',
  },
  QUEUE: {
    list: 'publishingQueueHearings',
    isFetching: false,
    params: { published: 'True' },
    formattedMessage: 'publishingQueue',
    iconName: 'calendar-check-o',
  },
  DRAFTS: {
    list: 'draftHearings',
    isFetching: false,
    params: { published: 'False' },
    formattedMessage: 'drafts',
    iconName: 'pencil-square-o',
  },
  OWN: {
    list: 'ownHearings',
    isFetching: false,
    params: {},
    formattedMessage: 'ownHearings',
    iconName: 'user',
    role: 'link',
  },
};

const AdminFilters = [HearingLists.PUBLISHED, HearingLists.QUEUE, HearingLists.DRAFTS, HearingLists.OWN];

const getHearingListParams = (listName) => {
  const defaultParams = { include: 'geojson' };
  const hearingList = find(HearingLists, ({ list }) => list === listName);
  const hearingListParams = get(hearingList, 'params', {});

  const params = { ...defaultParams, ...hearingListParams };

  switch (listName) {
    case HearingLists.PUBLISHED.list:
      params.open_at_lte = now();
      break;
    case HearingLists.QUEUE.list:
      params.open_at_gt = now();
      break;
    default:
    // Don't edit params further
  }
  return params;
};

function Hearings({
  user,
  labels,
  fetchInitialHearingList,
  fetchAllHearings,
  hearingLists,
  fetchMoreHearings,
  intl,
  language,
  fetchLabels,
}) {
  const params = useParams();
  const location = useLocation();
  const navigate = useNavigate();
  const [adminFilter, setAdminFilter] = useState(null);
  const [initHearingsFetched, setInitHearingsFetched] = useState(false);
  const [showOnlyOpen, setShowOnlyOpen] = useState(false);
  const [showOnlyClosed, setShowOnlyClosed] = useState(false);
  const [sortBy, setSortBy] = useState('-created_at');

  const tab = params.tab || 'list';

  const forwardToUserHearings = () => {
    const searchParams = parseQuery(location.search);

    navigate({
      pathname: '/user-hearings',
      search: stringifyQuery(searchParams),
    });
  };

  const getHearingListName = useCallback(
    () => (isAdmin(user) ? adminFilter : HearingLists.ALL.list),
    [adminFilter, user],
  );

  const getLabelsFromQuery = (labelsInQuery = []) => {
    if (Array.isArray(labelsInQuery)) return labelsInQuery;

    return [labelsInQuery];
  };

  const getSearchParams = useCallback(() => {
    const searchParams = {};
    if (parseQuery(location.search).search) Object.assign(searchParams, { title: parseQuery(location.search).search });
    if (parseQuery(location.search).label) {
      Object.assign(searchParams, { label: getLabelsFromQuery(parseQuery(location.search).label.toString()) });
    }
    return searchParams;
  }, [location.search]);

  const fetchHearingList = useCallback(() => {
    const list = getHearingListName();
    const filterByOpen = (showOnlyOpen && !showOnlyClosed) || (!showOnlyOpen && showOnlyClosed);
    const fetchParams = {
      ...getHearingListParams(list),
      ordering: sortBy,
      ...(filterByOpen && { open: showOnlyOpen }),
      ...getSearchParams(),
    };

    if (tab === 'map') {
      fetchAllHearings(list, fetchParams);
    } else {
      fetchInitialHearingList(list, fetchParams);
    }
    setInitHearingsFetched(true);
  }, [
    fetchAllHearings,
    fetchInitialHearingList,
    getHearingListName,
    getSearchParams,
    showOnlyClosed,
    showOnlyOpen,
    sortBy,
    tab,
  ]);

  const handleAdminFilter = (filter) => {
    if (filter === 'ownHearings') {
      forwardToUserHearings();
    } else {
      setAdminFilter(filter);
    }
  };

  const handleSearch = (searchTitle, force = false) => {
    const searchParams = parseQuery(location.search);
    const searchParamsEmpty = isEmpty(searchParams.search);
    const searchPhraseEmpty = isEmpty(searchTitle);
    const searchPhraseUpdated = searchParams.search !== searchTitle;

    // Don't do anything if:
    // search is not forced
    // AND
    // current search params are the same as the inputted search
    // OR
    // both search params and currently inputted search are empty
    if (!force && (searchParams.search === searchTitle || (searchParamsEmpty && searchPhraseEmpty))) {
      return;
    }

    if (searchPhraseEmpty) {
      delete searchParams.search;
    } else {
      searchParams.search = searchTitle;
    }

    if (searchPhraseUpdated || force) {
      navigate({
        path: location.path,
        search: stringifyQuery(searchParams),
      });
    }
  };

  const handleSelectLabels = (selectLabels) => {
    const selectArray = Array.isArray(selectLabels) ? selectLabels : [selectLabels];
    const searchParams = parseQuery(location.search);
    searchParams.label = selectArray.map(({ id }) => id);
    navigate({
      path: location.pathname,
      search: stringifyQuery(searchParams),
    });
  };

  const handleSort = (newSortBy, newShowOnlyOpen, newShowOnlyClosed) => {
    setSortBy(newSortBy);
    setShowOnlyOpen(newShowOnlyOpen);
    setShowOnlyClosed(newShowOnlyClosed);
    setInitHearingsFetched(false);
  };

  const getIsLoading = () => {
    const name = getHearingListName();
    return get(hearingLists, [name, 'isFetching'], true);
  };

  const getHearingsCount = () => {
    const list = getHearingListName();
    return get(hearingLists, [list, 'count'], 0);
  };

  const getHearings = () => {
    const hearingListKey = getHearingListName();
    return get(hearingLists, [hearingListKey, 'data'], []);
  };

  const handleReachBottom = () => {
    const list = getHearingListName();

    if (
      hearingLists[list] &&
      hearingLists[list].count > hearingLists[list].data.length &&
      typeof hearingLists[list].next === 'string' &&
      !hearingLists[list].isLoading
    ) {
      fetchMoreHearings(list);
    }
  };

  const toggleShowOnlyOpen = () => {
    setShowOnlyOpen(!showOnlyOpen);
  };

  const { formatMessage } = intl;

  const labelsInQuery = Array.isArray(parseQuery(location.search).label)
    ? parseQuery(location.search).label
    : [parseQuery(location.search).label];

  const selectedLabels = labels
    .filter((label) => includes(labelsInQuery, label.id.toString()))
    .map((label) => {
      const translatedLabel = getAttr(label.label, language);
      return {
        id: label.id,
        label: translatedLabel,
        value: translatedLabel
      }});

  const searchTitle = parseQuery(location.search).search;

  const hearings = getHearings();

  const hearingCount = getHearingsCount();

  useEffect(() => {
    fetchLabels();
    // Hearing List is fetched when labels are available -> componentWillReceiveProps
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (!initHearingsFetched && !isEmpty(labels)) {
      fetchHearingList();
    }
  }, [initHearingsFetched, fetchHearingList, labels]);

  useEffect(() => {
    const shouldSetAdminFilter = isAdmin(user) && (!user || !adminFilter);
    const shouldNullAdminFilter = !user || !isAdmin(user);
    const shouldFetchHearings = labels && (!user || (user && adminFilter)) && !isEmpty(getHearingListName());

    if (shouldSetAdminFilter) {
      handleAdminFilter(AdminFilters[0].list);
    }

    if (shouldNullAdminFilter) {
      setAdminFilter(null);
    }

    if (shouldFetchHearings) {
      fetchHearingList();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [user, labels, params, location, adminFilter]);

  if (user && user.isFetching) {
    return <LoadSpinner />;
  }

  return (
    <div className='hearings'>
      <section className='page-section page-section--all-hearings-header'>
        <div className='container'>
          <Row>
            <Col md={10} mdPush={1}>
              <Helmet
                title={formatMessage({ id: 'allHearings' })}
                meta={[
                  { name: 'description', content: formatMessage({ id: 'descriptionTag' }) },
                  { property: 'og:description', content: formatMessage({ id: 'descriptionTag' }) },
                ]}
              />
              <FormattedMessage id='allHearings'>{(txt) => <h1 className='page-title'>{txt}</h1>}</FormattedMessage>
              {isAdmin(user) && (
                <AdminFilterSelector
                  onSelect={handleAdminFilter}
                  options={AdminFilters}
                  active={getHearingListName()}
                />
              )}
              {isAdmin(user) && <CreateHearingButton to={{ path: '/hearing/new' }} />}
            </Col>
          </Row>
        </div>
      </section>
      {!isEmpty(labels) ? (
        <WrappedHearingList
          hearings={hearings}
          hearingCount={hearingCount}
          selectedLabels={selectedLabels ? [].concat(selectedLabels) : []}
          searchPhrase={searchTitle}
          isLoading={getIsLoading()}
          labels={labels}
          showOnlyOpen={showOnlyOpen}
          handleSort={handleSort}
          handleSearch={handleSearch}
          handleSelectLabels={handleSelectLabels}
          toggleShowOnlyOpen={toggleShowOnlyOpen}
          language={language}
          tab={tab}
          intl={intl}
          handleReachBottom={handleReachBottom}
          onTabChange={(value) => {
            const url = `/hearings/${value}`;
            navigate({
              pathname: url,
              search: location.search,
            });
          }}
        />
      ) : (
        <LoadSpinner />
      )}
    </div>
  );
}

Hearings.propTypes = {
  fetchInitialHearingList: PropTypes.func,
  fetchMoreHearings: PropTypes.func,
  fetchLabels: PropTypes.func,
  hearingLists: PropTypes.objectOf(
    PropTypes.shape({
      isFetching: PropTypes.bool,
      data: PropTypes.arrayOf(hearingShape),
    }),
  ),
  labels: PropTypes.arrayOf(labelShape),
  language: PropTypes.string,
  user: PropTypes.shape({
    data: userShape,
    isFetching: PropTypes.bool,
  }),
};

const mapStateToProps = (state) => ({
  hearingLists: state.hearingLists,
  labels: state.labels.data,
  language: state.language,
  user: getUser(state), // Using this on purpose without selector
});

const mapDispatchToProps = (dispatch) => ({
  fetchInitialHearingList: (listName, params) =>
    dispatch(Actions.fetchInitialHearingList(listName, '/v1/hearing/', params)),
  fetchAllHearings: (list, params) => dispatch(Actions.fetchHearingList(list, '/v1/hearing', params)),
  fetchMoreHearings: (listName) => dispatch(Actions.fetchMoreHearings(listName)),
  fetchLabels: () => dispatch(Actions.fetchLabels()),
});

export const UnconnectedHearings = Hearings;

export default connect(mapStateToProps, mapDispatchToProps)(injectIntl(Hearings));
