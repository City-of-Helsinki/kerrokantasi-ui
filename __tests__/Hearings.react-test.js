import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import {Hearings} from '../src/views/Hearings/index';
import {MemoryRouter} from 'react-router-dom';

test('Hearings component should render as expected', () => {
  const props = {
    intl: {
      formatMessage: () => 'translatedText',
    },
    dispatch: () => {},
    fetchLabels: () => {},
    match: {
      params: {
        tab: 'list'
      }
    },
    location: {
      search: ''
    },
    hearingLists: {
      allHearings: {
        data: []
      }
    },
    user: {}
  }

  const component = shallow(<MemoryRouter><Hearings {...props} /></MemoryRouter>).dive().dive();
  const tree = toJson(component);
  console.log(tree);
  expect(tree).toMatchSnapshot();
});