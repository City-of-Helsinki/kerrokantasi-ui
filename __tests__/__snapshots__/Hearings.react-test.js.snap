exports[`test Hearings component should render as expected 1`] = `
<div
  className="hearings">
  <section
    className="page-section page-section--all-hearings-header">
    <div
      className="container">
      <Row
        bsClass="row"
        componentClass="div">
        <Col
          bsClass="col"
          componentClass="div"
          md={10}
          mdPush={1}>
          <HelmetWrapper
            title="translatedText" />
          <h1
            className="page-title">
            <FormattedMessage
              id="allHearings"
              tagName="span"
              values={Object {}} />
          </h1>
        </Col>
      </Row>
    </div>
  </section>
  <InjectIntl(HearingList)
    handleReachBottom={[Function]}
    handleSearch={[Function]}
    handleSelectLabels={[Function]}
    handleSort={[Function]}
    hearings={Array []}
    isLoading={true}
    onTabChange={[Function]}
    selectedLabels={Array []}
    showOnlyOpen={false}
    tab="list"
    toggleShowOnlyOpen={[Function]} />
</div>
`;
